### PR TITLE
feat: sort policies in alphabetical order before executing

### PR DESCRIPTION
**Pull request by vishal-chdhry**
## Explanation

Maintains a sorted list of policies, and inserts to the sorted list.

Should be faster than sorting in `CompiledPolicies(context.Context) ([]PolicyFunc, error)` as policies are fetched more than they are created or updated

TODO: Chainsaw tests

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.

### CODE REVIEW

File: pkg/policy/provider.go [hunk 1]
```diff
@@ -28,39 +30,66 @@ type policyReconciler struct {
...
+	} else {
+		r.policies = slices.Insert(r.policies, i, pol)
+	}
+}
+
+func (r *policyReconciler) deletePolicy(name types.NamespacedName) {
+	deletecmp := func(p policy) bool {
+		return name == p.name
+	}
+	r.policies = slices.DeleteFunc(r.policies, deletecmp)
```
Comment (eddycharly) [inline pkg/policy/provider.go:57]: Complexity on delete is high (O(n))

File: pkg/policy/provider.go [hunk 2]
```diff
@@ -28,39 +30,66 @@ type policyReconciler struct {
...
+	name       types.NamespacedName
+	policyFunc PolicyFunc
+}
+
+func (r *policyReconciler) addPolicy(pol policy) {
+	cmp := func(current, target policy) int {
+		return strings.Compare(current.name.String(), target.name.String())
+	}
+
+	if i, found := slices.BinarySearchFunc(r.policies, pol, cmp); found {
```
Comment (eddycharly) [inline pkg/policy/provider.go:46]: Complexity on find in O(log n)

File: pkg/policy/provider.go [hunk 3]
```diff
@@ -40,13 +43,35 @@ func newPolicyReconciler(client client.Client, compiler Compiler) *policyReconci
 	}
 }
 
+func (r *policyReconciler) setSortPoliciesFunc() func() []PolicyFunc {
+	return sync.OnceValue(func() []PolicyFunc {
+		keys := make([]types.NamespacedName, 0, len(r.policies))
+		for k := range r.policies {
+			keys = append(keys, k)
+		}
```
Comment (eddycharly) [inline pkg/policy/provider.go:51]:
```suggestion
		keys := maps.Keys(r.policies)
```

File: pkg/policy/provider.go [hunk 4]
```diff
@@ -40,13 +43,35 @@ func newPolicyReconciler(client client.Client, compiler Compiler) *policyReconci
 	}
 }
 
+func (r *policyReconciler) setSortPoliciesFunc() func() []PolicyFunc {
```
Comment (eddycharly) [inline pkg/policy/provider.go:46]:
```suggestion
func setSortPoliciesFunc(policies map[types.NamespacedName]PolicyFunc) func() []PolicyFunc {
```

Comment (vishal-chdhry) [reply pkg/policy/provider.go:46]: Oh so they are passed by reference, I had the doubt that the state might be stale if I pass it as argument

File: pkg/policy/provider.go [hunk 5]
```diff
@@ -25,28 +26,44 @@ func NewKubeProvider(mgr ctrl.Manager, compiler Compiler) (Provider, error) {
...
+	return sync.OnceValue(func() []PolicyFunc {
+		keys := maps.Keys(policies)
+
+		slices.Sort(keys)
+
+		out := make([]PolicyFunc, 0, len(policies))
+		for _, key := range keys {
+			out = append(out, policies[key])
+		}
+		return out
```
Comment (eddycharly) [inline pkg/policy/provider.go:55]: Think about this func, something important is missing.

File: pkg/policy/provider.go [hunk 6]
```diff
@@ -25,28 +26,44 @@ func NewKubeProvider(mgr ctrl.Manager, compiler Compiler) (Provider, error) {
...
 	return &policyReconciler{
 		client:   client,
 		compiler: compiler,
 		lock:     &sync.RWMutex{},
-		policies: map[types.NamespacedName]PolicyFunc{},
+		policies: map[string]PolicyFunc{},
 	}
 }
 
+func setSortPoliciesFunc(policies map[string]PolicyFunc) func() []PolicyFunc {
```
Comment (eddycharly) [inline pkg/policy/provider.go:45]: Now, this can probably be made generic, key is comparable.

File: pkg/policy/provider.go [hunk 7]
```diff
@@ -25,28 +26,44 @@ func NewKubeProvider(mgr ctrl.Manager, compiler Compiler) (Provider, error) {
...
+	return sync.OnceValue(func() []PolicyFunc {
+		keys := maps.Keys(policies)
+
+		slices.Sort(keys)
```
Comment (eddycharly) [inline pkg/policy/provider.go:49]:
```suggestion
		keys := slices.Sorted(maps.Keys(policies))
```

File: pkg/policy/provider.go [hunk 8]
```diff
@@ -25,28 +27,41 @@ func NewKubeProvider(mgr ctrl.Manager, compiler Compiler) (Provider, error) {
...
 	return &policyReconciler{
 		client:   client,
 		compiler: compiler,
 		lock:     &sync.RWMutex{},
-		policies: map[types.NamespacedName]PolicyFunc{},
+		policies: map[string]PolicyFunc{},
 	}
 }
 
+func setSortPoliciesFunc[K cmp.Ordered](policies map[K]PolicyFunc) func() []PolicyFunc {
```
Comment (eddycharly) [inline pkg/policy/provider.go:46]: How about the map value ?

File: pkg/policy/provider.go [hunk 9]
```diff
@@ -25,28 +27,50 @@ func NewKubeProvider(mgr ctrl.Manager, compiler Compiler) (Provider, error) {
...
+		out = append(out, in[key])
+	}
+	return out
+}
+
 func (r *policyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	var policy v1alpha1.AuthorizationPolicy
+
+	// Reset the sorted func on every reconcile so the policies get resorted in next call
+	defer func() {
```
Comment (eddycharly) [inline pkg/policy/provider.go:61]: I don't think we want to do that in all cases.

Comment (eddycharly) [reply pkg/policy/provider.go:61]: And... it's not thread safe ;-)

File: pkg/policy/provider.go [hunk 10]
```diff
@@ -60,16 +84,11 @@ func (r *policyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
 }
 
 func (r *policyReconciler) CompiledPolicies(ctx context.Context) ([]PolicyFunc, error) {
-	r.lock.RLock()
-	defer r.lock.RUnlock()
-	out := make([]PolicyFunc, 0, len(r.policies))
-	for _, policy := range r.policies {
-		out = append(out, policy)
-	}
+slices.Clone(r.sortPolicies())
```
Comment (eddycharly) [inline pkg/policy/provider.go:92]:
```suggestion
return slices.Clone(r.sortPolicies())
```

**Review by eddycharly — APPROVED** [review]

File: pkg/policy/provider.go [hunk 11]
```diff
@@ -25,28 +27,51 @@ func NewKubeProvider(mgr ctrl.Manager, compiler Compiler) (Provider, error) {
...
 type policyReconciler struct {
-	client   client.Client
-	compiler Compiler
-	lock     *sync.RWMutex
-	policies map[types.NamespacedName]PolicyFunc
+	client       client.Client
+	compiler     Compiler
+	lock         *sync.Mutex
+	policies     map[string]PolicyFunc
+	sortPolicies func() []PolicyFunc
```
Comment (eddycharly) [inline pkg/policy/provider.go:34]: Don't you need to initialize this ?

File: pkg/policy/provider.go [hunk 12]
```diff
@@ -60,16 +85,11 @@ func (r *policyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
 
 func (r *policyReconciler) CompiledPolicies(ctx context.Context) ([]PolicyFunc, error) {
-	r.lock.RLock()
-	defer r.lock.RUnlock()
-	out := make([]PolicyFunc, 0, len(r.policies))
-	for _, policy := range r.policies {
-		out = append(out, policy)
-	}
-	return out, nil
+	return slices.Clone(r.sortPolicies()), nil
```
Comment (eddycharly) [inline pkg/policy/provider.go:94]: `r.sortPolicies` can be nil here right ?

File: pkg/policy/provider.go [hunk 13]
```diff
@@ -25,28 +27,54 @@ func NewKubeProvider(mgr ctrl.Manager, compiler Compiler) (Provider, error) {
...
 func newPolicyReconciler(client client.Client, compiler Compiler) *policyReconciler {
 	return &policyReconciler{
 		client:   client,
 		compiler: compiler,
-		lock:     &sync.RWMutex{},
-		policies: map[types.NamespacedName]PolicyFunc{},
+		lock:     &sync.Mutex{},
+		policies: map[string]PolicyFunc{},
+		sortPolicies: func() []PolicyFunc {
+			return []PolicyFunc{}
```
Comment (eddycharly) [inline pkg/policy/provider.go:44]:
```suggestion
			return nil
```

**Review by eddycharly — APPROVED** [review]

**Pull request merged**